### DISCUSSION TITLE
I need more powerful scripting capabilities

### DISCUSSION THREAD

**Discussion by carboncope** [Hekili/hekili#1583] [Q&A]
Is there anyway to bypass the simc parsing for criteria and tell hekili to loadscript a pure lua in the criteria box?

What I am trying to do:

1. I turn on custom1 toggle to trigger covenant abilities
2. I want the custom1 toggle to auto turn off after a covenant.cooldown >0 is triggered.
3. In my mind I would have a action that has criteria that can run a pure lua script with a modifier comment (-- purelua ) or something to let hekili know hey dont try to parse this as simc. Let the user return true or false.
4. Even better would be to have a new action type [script] similar to [variable] that does not try to parse simc
5. 
```

-- purelua

if (getglobal("Hekili").DB.profile.toggles.custom1.value and cooldown.covenantspell.remains > 60 ) then

	getglobal("Hekili").DB.profile.toggles.custom1.value  = false -- turn custom1 off

end

return false -- let the user decide to return true or false for criteria
```

**Comment by Hekili** [comment]
I don't intend to support this in this way, simply because it creates a maintenance nightmare in terms of figuring out what someone did with a script when they report a bug report.

I'm not sure why you want custom1 to be your toggle for Covenants when there's already a covenants toggle.

I could potentially make it so that using an ability will turn a toggle on/off, and let you control that via /hekili > Abilities.

**Reply by carboncope** [reply]
> I don't intend to support this in this way, simply because it creates a maintenance nightmare in terms of figuring out what someone did with a script when they report a bug report.

The answer is dont support it. If you get a bug report with --purelua in it, quick reply we dont support bug reports for that. Would be super useful to have it though.

> I'm not sure why you want custom1 to be your toggle for Covenants when there's already a covenants toggle.

Just giving example for context.

> I could potentially make it so that using an ability will turn a toggle on/off, and let you control that via /hekili > Abilities.

I'll accept that, but [script] type without bug support would be way more awesome.

**Reply by dubudevs** [reply]
My 2c; custom APLs and modified APLs are also unsupported but are an incredible tool to have available for the few people who want to use it. I think this would fall into that category too. It doesn't need to be documented anywhere other than a hover on a box or something, but for the people who can make use of it there's nothing like it (presuming its a relatively simple thing to implement).

**Reply by Hekili** [reply]
Custom APLs are always *allowed* but unsupported, and that will remain true.

One of the issues with freely allowing Lua scripts is that it is **not** intuitive, dealing with "is this true now, or is this true after multiple other abilities have been used?"  If you write code that checks for an aura yourself (i.e., by ID which I don't have supported at this time), maybe your next action would cause that aura to be consumed, but the predictive engine would still expect it to be there and give a misleading recommendation.

Unfortunately, having a feature that is "unsupported" still results in a lot of support being required later on.  I still lean toward "let me know what you're wanting to do, and maybe I can implement it" vs. "go hog wild in the addon environment."

**Reply by carboncope** [reply]
Thanks for the replies @Hekili ,

The toggle feature would be pretty nice, if you have time.

If you add that option to Abilities, can you also add it to Gear and Items options?

(New Dropdown option)
**Cooldown Flips Toggle:**
None
Cooldowns
Covenants
Interrupts
Defensives
Potions
Custom1
Custom2
